File: src/546_Remove_Boxes.cpp
// 546. Remove Boxes
// https://leetcode.cn/problems/remove-boxes/description/


#include<cstdio>
#include<cstring>
#include<algorithm>
#include<iostream>
#include<string>
#include<vector>
#include<stack>
#include<bitset>
#include<cstdlib>
#include<cmath>
#include<set>
#include<list>
#include<deque>
#include<map>
#include<queue>
using namespace std;
typedef long long ll;
const double PI = acos(-1.0);
const double eps = 1e-6;
const int INF = 0x3f3f3f3f;

class Solution {
public:
    // dp数组用于存储子问题的结果。维度是100x100x100，用于存储三个参数的子问题。
    int dp[100][100][100];

    int removeBoxes(vector<int>& boxes) {
        // 将dp数组初始化为0
        memset(dp, 0, sizeof dp);
        // 调用辅助函数calculatePoints来计算最大分数
        return calculatePoints(boxes, 0, boxes.size() - 1, 0);
    }

    int calculatePoints(vector<int>& boxes, int l, int r, int k) {
        // 基础情况：如果左边界大于右边界，则返回0
        if (l > r) {
            return 0;
        }

        // 如果当前子问题的解还未被计算
        if (dp[l][r][k] == 0) {
            // 默认情况：移除右边界的盒子并获得(k+1)*(k+1)的分数
            dp[l][r][k] = calculatePoints(boxes, l, r - 1, 0) + (k + 1) * (k + 1);

            // 遍历当前子数组，寻找和右边界盒子颜色相同的盒子，看是否有机会合并
            for (int i = l; i < r; i++) {
                if (boxes[i] == boxes[r]) {
                    // 尝试合并和右边界相同颜色的盒子，并计算两个子问题：一个是[l, i]范围内的盒子，另一个是[i+1, r-1]范围内的盒子
                    dp[l][r][k] = max(dp[l][r][k], calculatePoints(boxes, l, i, k + 1) + calculatePoints(boxes, i + 1, r - 1, 0));
                }
            }
        }
        // 返回当前子问题的解
        return dp[l][r][k];
    }
};

File: src/546_Remove_Boxes.py
#  546_Remove_Boxes
#  https://leetcode.cn/problems/remove-boxes/

# 给出一些不同颜色的盒子 boxes ，盒子的颜色由不同的正数表示。

# 你将经过若干轮操作去去掉盒子，直到所有的盒子都去掉为止。
# 每一轮你可以移除具有相同颜色的连续 k 个盒子（k >= 1），这样一轮之后你将得到 k * k 个积分。

# 返回 你能获得的最大积分和 。

# 示例 1：
# 输入：boxes = [1,3,2,2,2,3,4,3,1]
# 输出：23
# 解释：
# [1, 3, 2, 2, 2, 3, 4, 3, 1] 
# ----> [1, 3, 3, 4, 3, 1] (3*3=9 分) 
# ----> [1, 3, 3, 3, 1] (1*1=1 分) 
# ----> [1, 1] (3*3=9 分) 
# ----> [] (2*2=4 分)

# 示例 2：
# 输入：boxes = [1,1,1]
# 输出：9

# 示例 3：
# 输入：boxes = [1]
# 输出：1

from typing import List

class Solution:
    def removeBoxes(self, boxes: List[int]) -> int:
        dp = [[[0] * 100 for _ in range(100)] for _ in range(100)]
        
        def calculatePoints(boxes, l, r, k):
            if l > r:
                return 0
            
            if dp[l][r][k] == 0:
                dp[l][r][k] = calculatePoints(boxes, l, r - 1, 0) + (k + 1) * (k + 1)
            
                for i in range(l, r):
                    if boxes[i] == boxes[r]:
                        dp[l][r][k] = max(dp[l][r][k], calculatePoints(boxes, l, i, k+1) + calculatePoints(boxes, i+1, r-1, 0))
                        
            return dp[l][r][k]
            
        
        calculatePoints(boxes, 0, len(boxes) - 1, 0)
        return dp[0][len(boxes)-1][0]    
        
        
        
        
s = Solution()
ans = s.removeBoxes([1,3,2,2,2,3,4,3,1])
print(ans)        


s = Solution()
ans = s.removeBoxes([1,1,1])
print(ans)


s = Solution()
ans = s.removeBoxes([1])
print(ans)